compiler/noirc_frontend/src/hir/type_check/expr.rs

```diff
@@ -1159,11 +1160,29 @@ impl<'interner> TypeChecker<'interner> {
                 if let TypeBinding::Bound(binding) = &*int.borrow() {
                     return self.infix_operand_type_rules(binding, op, other, span);
                 }
-
+                if op.kind == BinaryOpKind::ShiftLeft || op.kind == BinaryOpKind::ShiftRight {
+                    self.unify(
+                        rhs_type,
+                        &Type::Integer(Signedness::Unsigned, IntegerBitSize::Eight),
+                        || TypeCheckError::TypeMismatchWithSource {
+                            expected: Type::Integer(Signedness::Unsigned, IntegerBitSize::Eight),
+                            actual: rhs_type.clone(),
+                            source: Source::Binary,
+                            span,
+                        },
+                    );
+                    return Ok((lhs_type.clone(), true));
+                }
```

Comment:
`lhs` still needs to be bound to an integer type variable here. That's why you're getting inference errors in tests later on where you now need to cast `64 as u32` specifically instead of it being inferred. Also - shouldn't this be the new error variant that was just added?

Reply:
For the error message, I wonder if I should not always use the type mismatch error instead of the new variant? (i.e I remove the new variant)
For `lhs`, I feel weird to bind it to `rhs` now that `rhs` must always be a u8.

What do you think?

Reply:
I think since we already have the new error type we might as well use it.

The `lhs` shouldn't be bound to the `rhs` since they're not the same types anymore but it'll still need to be bound to a polymorphic integer type variable.

Reply:
I have changed the error type.
For the binding, I don't think it is correct to bind it to a polymorphic integer since this would fail when shifting a U128 for instance.

Reply:
I was thinking we'd copy the existing check used elsewhere to only apply it if it is numeric:
```rs
        if lhs_type.is_numeric() {
```
This way we can still narrow an integer literal from `IntegerOrField` to just `Integer`

Reply:
Yes that makes sense, I've done it.